pytorch_lightning/strategies/bagua.py

```diff
@@ -0,0 +1,239 @@
+from pytorch_lightning.plugins.io.checkpoint_plugin import CheckpointIO
+from pytorch_lightning.plugins.precision import PrecisionPlugin
+from pytorch_lightning.strategies.ddp import DDPStrategy
+from pytorch_lightning.strategies.strategy import TBroadcast
+from pytorch_lightning.utilities.distributed import ReduceOp
+from pytorch_lightning.utilities.enums import _StrategyType
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from pytorch_lightning.utilities.imports import _BAGUA_AVAILABLE
+from pytorch_lightning.utilities.seed import reset_seed
+
+if _BAGUA_AVAILABLE:
+    import bagua.torch_api as bagua
+    from bagua.torch_api.algorithms import Algorithm
+    from bagua.torch_api.algorithms.q_adam import QAdamOptimizer
+    from bagua.torch_api.communication import allreduce_inplace, barrier, broadcast_object, is_initialized
+    from bagua.torch_api.communication import ReduceOp as BaguaReduceOp
+    from bagua.torch_api.data_parallel.distributed import DistributedDataParallel_V1_9_0 as BaguaDistributedDataParallel
+else:
+    BaguaReduceOp = None
+    BaguaDistributedDataParallel = None
+
+log = logging.getLogger(__name__)
+
+
+class LightningBaguaModule(_LightningModuleWrapperBase):
+    def __init__(self, pl_module: "pl.LightningModule") -> None:
+        super().__init__(pl_module)
+        # Bagua use `bagua_module_name` to distinguish different modules
+        self._bagua_module_name = pl_module._get_name() + str(id(pl_module))
+
+
+_bagua_reduce_ops: Dict[ReduceOp, BaguaReduceOp] = {}
+
+
+def _from_torch_reduce_op(op: ReduceOp) -> Optional[BaguaReduceOp]:
+    """Convert a `torch.distributed.ReduceOp` to its equivalent `bagua.torch_api.ReduceOp`."""
+    global _bagua_reduce_ops
+
+    if len(_bagua_reduce_ops) == 0:
+        _bagua_reduce_ops.update(
+            {
+                ReduceOp.SUM: BaguaReduceOp.SUM,
+                ReduceOp.PRODUCT: BaguaReduceOp.PRODUCT,
+                ReduceOp.MIN: BaguaReduceOp.MIN,
+                ReduceOp.MAX: BaguaReduceOp.MAX,
+                ReduceOp.BAND: BaguaReduceOp.BAND,
+                ReduceOp.BOR: BaguaReduceOp.BOR,
+                ReduceOp.BXOR: BaguaReduceOp.BXOR,
+            }
+        )
+
+    return _bagua_reduce_ops.get(op, None)
+
+
+class BaguaStrategy(DDPStrategy):
+    distributed_backend = _StrategyType.BAGUA
+
+    def __init__(
+        self,
+        algorithm: str = "gradient_allreduce",
+        do_flatten: bool = True,
+        accelerator: Optional["pl.accelerators.accelerator.Accelerator"] = None,
+        parallel_devices: Optional[List[torch.device]] = None,
+        cluster_environment: Optional[ClusterEnvironment] = None,
+        checkpoint_io: Optional[CheckpointIO] = None,
+        precision_plugin: Optional[PrecisionPlugin] = None,
+        **bagua_kwargs: Union[Any, Dict[str, Any]],
+    ):
+        """Strategy for training using the `Bagua <https://github.com/BaguaSys/bagua>`_ library, with advanced
```

Comment:
Can Bagua run on CPU at all? For debugging for example.

Reply:
Discussed with @wangraying 
No, it does not support GPU. Added a sentence in the docstring.